src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/TokenBucketRateLimiter.cs

```diff
@@ -0,0 +1,326 @@
+                TokenBucketRateLimiterOptions options = _options;
+                int maxPermits = options.TokenLimit;
+                int resourcesToAdd;
+
+                if (availablePermits < maxPermits)
+                {
+                    resourcesToAdd = Math.Min(options.TokensPerPeriod, maxPermits - availablePermits);
+                }
+                else
+                {
+                    // All tokens available, nothing to do
+                    return;
+                }
+
+                // Process queued requests
+                Deque<RequestRegistration> queue = _queue;
+
+                _tokenCount += resourcesToAdd;
+                Debug.Assert(_tokenCount <= _options.TokenLimit);
+                while (queue.Count > 0)
+                {
+                    RequestRegistration nextPendingRequest =
+                          options.QueueProcessingOrder == QueueProcessingOrder.OldestFirst
+                          ? queue.PeekHead()
+                          : queue.PeekTail();
+
+                    if (_tokenCount >= nextPendingRequest.Count)
+                    {
+                        // Request can be fulfilled
+                        nextPendingRequest =
+                            options.QueueProcessingOrder == QueueProcessingOrder.OldestFirst
+                            ? queue.DequeueHead()
+                            : queue.DequeueTail();
+
+                        _queueCount -= nextPendingRequest.Count;
+                        _tokenCount -= nextPendingRequest.Count;
+                        Debug.Assert(_queueCount >= 0);
+                        Debug.Assert(_tokenCount >= 0);
+
+                        if (!nextPendingRequest.Tcs.TrySetResult(SuccessfulLease))
+                        {
+                            // Queued item was canceled so add count back
+                            _tokenCount += nextPendingRequest.Count;
+                        }
+                        nextPendingRequest.CancellationTokenRegistration.Dispose();
+                    }
+                    else
+                    {
+                        // Request cannot be fulfilled
+                        break;
+                    }
+                }
+            }
+        }
+
+        private class TokenBucketLease : RateLimitLease
+        {
+            private readonly TimeSpan? _retryAfter;
+
+            public TokenBucketLease(bool isAcquired, TimeSpan? retryAfter)
+            {
+                IsAcquired = isAcquired;
+                _retryAfter = retryAfter;
+            }
+
+            public override bool IsAcquired { get; }
+
+            public override IEnumerable<string> MetadataNames => Enumerable();
+
+            private IEnumerable<string> Enumerable()
+            {
+                if (_retryAfter is not null)
+                {
+                    yield return MetadataName.RetryAfter.Name;
+                }
+            }
```

Comment:
Similar comment to earlier.  We can make MetadataNames cheaper and less code; not sure how impactful that actually is.